### ISSUE TITLE
error: 32-bit absolute addressing is not supported in 64-bit mode

### ISSUE THREAD

**Issue by xhliu** [falgon/htcc#7]
I was following the README, but got the following error. Does it work on Mac OS?
```
> gcc -no-pie t.s -o out
clang: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
t.s:11:2: error: 32-bit absolute addressing is not supported in 64-bit mode
 push offset .L.data.0
```

```
gcc  --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by falgon** [comment]
@xhliu 
Thank you for trying! This tiny compiler is currently intended to work on Linux and does not support working on macOS (These are incompatible in small details such as static links).
I am currently working on the task (https://github.com/falgon/htcc/issues/2) and plan to start the task of outputting the LLVM IR code after this is complete, so in the future macOS may also be supported.

**Comment by xhliu** [comment]
@falgon 
How come there is no explicit array subscript operator `[]` in the parser? Is it implicitly converted to dereference operator `*`?

Also, how are multidimensional arrays (like `a[2][3][3]`) & nested structs (like `a.b.c`) supported?


BTW, the following `hello.c` does not compile
```
int printf();

struct A
{
    int a;
};

struct B
{
    int c;
    struct A b;
};

int main()
{
    struct A st = {42};
    struct B st2 = {12, st};
    printf("%d\n", st2.b.a);
}
```
```
> stack exec htcc -- hello.c > hello.s
hello.c:17:21: error: expected { initializer-list } or { initializer-list , }
    17 |     struct B st2 = {12, st};
       |                     ^~
```

**Comment by falgon** [comment]
@xhliu 
There is a parser for array subscript operators. For example [here](https://github.com/falgon/htcc/blob/master/src/Htcc/Parser/Parsing/Core.hs#L351).
But there is no `[]` operator for array access in C. The C `[]` is just a simple notation for accessing the elements of an array via pointers. An array is implicitly converted to a pointer to the first element of the array, except when used as an operand of `sizeof` or the unary `&`.

> BTW, the following hello.c does not compile

Wow, it supports nested structs, but it seems I couldn't test for initialization.
As shown below, it is possible to compile at present if it is assigned instead of initialized.
```c
int printf(); 
struct A 
{ 
    int a; 
}; 

struct B 
{ 
    int c; 
    struct A b; 
}; 

int main() 
{ 
    struct A st = { 42 }; 
    struct B st2; 
    st2.c = 12; 
    st2.b = st;
    printf("%d", st2.b.a); 
}
```
However, this constraint is just an implementation omission and will be fixed as task #2 progresses. 
Thank you for discovering the implementation omission.
